### PR TITLE
Merge twitter monitor contacts based on name

### PR DESCRIPTION
**Pull request by Woeler**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? |  X
| Related user documentation PR URL | Will come soon
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Added a configurable option to Twitter social monitoring. When turned on Mautic will try to match people found on Twitter with already existing contacts based on the name of the twitter account. This avoids having people double in your contacts list (one raw, and one twitter) plus it's an extra feature to the social monitoring in general.

Hubspot has a similar feature. Do note that this is never 100% accurate, but it should be close enough to be very useful for most users. Hence the reason why I made it configurable.

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
N/A

#### Steps to test this PR:
1. Create a social monitor (Twitter hashtag or mention)
2. Save
3. Set 'Match contact names' to 'Yes'
4. Start the Cronjob
5. See the magic happen

Note that in order for this to work you need a contact and a twitter account with the same name. i.e. contact name 'John Doe' and twitter account '@johndoe' with the name set as 'John Doe'. It will compare the contact name with the name set on twitter (so not the @ name).

#### List deprecations along with the new alternative:
N/A

#### List backwards compatibility breaks:
N/A

### CODE REVIEW

**Comment by escopecz** [comment]
Please fix the code standards: https://travis-ci.org/mautic/mautic/jobs/247565512#L597

**Comment by escopecz** [comment]
I take it back. The issue was in the staging branch. It is resolved now. Please merge the latest staging into this branch. Also, please revert back the file permissions.

**Comment by Woeler** [comment]
I'm a little new to this whole merging in GIT. What are the steps for doing this. I'd like to know the correct way so I have it for future reference. Thanks :)

**Comment by escopecz** [comment]
Here is how I'd do it, but there may be more ways to do it.

1. Go to `staging` branch: `git checkout staging`
2. Pull the latest changes from the official repo. I called it `upstream` on my local, so change it to whatever you call it: `git pull upstream staging`.
    2.1 If you don't have the original repository in your remotes yet, do this: `git remote add upstream git@github.com:mautic/mautic.git`
3. Go to your feature branch now `git checkout "Feature/Twitter_Monitor_Name_Matching"`. I'm adding the quotes here because the branch name contains the slash which confuses git on my local.
4. Merge the current staging to your branch: `git merge staging`
5. Push to your remote. I call it `origin` on my local, so change the name to yours: `git push origin/"Feature/Twitter_Monitor_Name_Matching"`

_I didn't test the commands I posted above, so there may be some issue, but it should show you the workflow_

**Comment by Woeler** [comment]
Thank you. Got it done. Travis passed.

**Comment by escopecz** [comment]
Cool! 

The last issue I see before testing is that the files have changed permission. You can see it in the [Files Changed](https://github.com/mautic/mautic/pull/4312/files) tab. For example:
`plugins/MauticSocialBundle/Command/MonitorTwitterBaseCommand.php 100644 → 100755`

The permissions for files must be 100644, not 100755. Could you change them back, please?

**Comment by Woeler** [comment]
I assume that's just a sudo chmod -R 644 myproject/ ?

**Comment by Woeler** [comment]
And done :)

**Comment by gingerling** [comment]
I will have a bash at the docs for this next week, if you are okay to wait

**Comment by mqueme** [comment]
this merged or updated the existing contact, and this message was displayed.

Updating existing lead ID #2546 (username)

but I don't know where to `Set 'Match contact names' to 'Yes'`?

File: plugins/MauticSocialBundle/Command/MonitorTwitterBaseCommand.php [hunk 1]
```diff
@@ -245,6 +253,38 @@ protected function createLeadsFromStatuses($statusList, $monitor)
                     ],
                 ]
             );
+            if ($monitorProperties['matchnames'] === 1) {
```
Comment (virlatinus) [inline plugins/MauticSocialBundle/Command/MonitorTwitterBaseCommand.php]: This line checks for a property called `matchnames`, but in the DB, the property is stored as `checknames`:
```
'a:2:{s:7:"hashtag";s:9:"TraficoGT";s:10:"checknames";i:1;}'
```

**Comment by virlatinus** [comment]
There is a couple of issues with this PR, the dropdown to select whether to match contact names appears only after saving a new social monitor or when editing an existing one:

image

Also, MonitorTwitterBaseCommand.php:256 checks for a property called `matchnames`:
```
if ($monitorProperties['matchnames'] === 1) {
```
but in the DB it is stored as `checknames`:
```
'a:2:{s:7:"hashtag";s:9:"TraficoGT";s:10:"checknames";i:1;}'
```

**Comment by dongilbert** [comment]
Since this hasn't been updated and I can't find the branch to push updates to, I'm closing this in favor of #5342

**Pull request closed without merging**